R/rvar-cast.R

```diff
@@ -135,118 +171,300 @@ invalidate_rvar_cache = function(x) {
+  })
+  out <- bind_rvars(rvars, arg_exprs = NULL, deparse.level = 0, axis = 1)
+
+  if (all(lengths(lapply(groups, function(x) dim(x[[1]]$draws))) <= 2)) {
+    # input was a bunch of vectors, ensure output is also a vector
+    .dimnames <- dimnames(out)
+    dim(out) <- length(out)
+    dimnames(out) <- .dimnames[1]
   }
-  # determine the number of chains
-  nchains_or_null <- lapply(x, function(x) if (dim(x)[[2]] %||% 1 == 1) NULL else attr(x, "nchains"))
-  .nchains <- Reduce(nchains2_common, nchains_or_null) %||% 1L
-
-  out <- new_rvar(.draws, .nchains = .nchains)
 
   # since we've already spent time calculating it, save the proxy in the cache
-  attr(out, "cache")$vec_proxy <- x
+  # - but only if the proxy only has one group (else we'd have to recalculate
+  # the bind above again, which is usually more expensive than generating the
+  # proxy itself)
+  if (length(groups) == 1) {
+    attr(out, "cache")$vec_proxy <- x
+  }
 
   out
 }
 
+#' @export
+vec_restore.rvar_factor = function(x, to, ...) {
+  x[lengths(x) == 0] <- make_rvar_proxy(rvar_factor(NA_integer_))
+  vec_restore.rvar(x, ...)
+}
+
+#' @export
+vec_restore.rvar_ordered = function(x, to, ...) {
+  x[lengths(x) == 0] <- make_rvar_proxy(rvar_ordered(NA_integer_))
+  vec_restore.rvar(x, ...)
+}
+
 
-# vec_ptype performance generic -------------------------------------------
+# vec_ptype performance generics -------------------------------------------
 
 #' @importFrom vctrs vec_ptype
 #' @export
 vec_ptype.rvar <- function(x, ..., x_arg = "") new_rvar()
+#' @export
+vec_ptype.rvar_factor <- function(x, ..., x_arg = "") new_rvar(factor())
+#' @export
+vec_ptype.rvar_ordered <- function(x, ..., x_arg = "") new_rvar(ordered(NULL, levels = levels(x)))
 
 
-# identity cast -----------------------------------------------------------
+# identity casts -----------------------------------------------------------
 
 #' @importFrom vctrs vec_ptype2
+#' @importFrom vctrs vec_cast
 #' @export
 vec_ptype2.rvar.rvar <- function(x, y, ...) new_rvar()
-
-#' @importFrom vctrs vec_cast
 #' @export
 vec_cast.rvar.rvar <- function(x, to, ...) x
 
+#' @export
+vec_ptype2.rvar_factor.rvar_factor <- function(x, y, ...) new_rvar(factor())
+#' @export
+vec_cast.rvar_factor.rvar_factor <- function(x, to, ...) x
+
+#' @export
+vec_ptype2.rvar_ordered.rvar_ordered <- function(x, y, ...) new_rvar(ordered(NULL))
+#' @export
+vec_cast.rvar_ordered.rvar_ordered <- function(x, to, ...) x
+
 
 # numeric and logical casts -----------------------------------------------
 
+# double -> rvar
 #' @export
 vec_ptype2.double.rvar <- function(x, y, ...) new_rvar()
-
 #' @export
 vec_ptype2.rvar.double <- function(x, y, ...) new_rvar()
-
 #' @export
 vec_cast.rvar.double <- function(x, to, ...) new_constant_rvar(x)
 
+# double -> rvar_factor
 #' @export
-vec_ptype2.integer.rvar <- function(x, y, ...) new_rvar()
+vec_cast.rvar_factor.double <- function(x, to, ...) new_constant_rvar(while_preserving_dims(as.factor, x))
```

Comment:
I suggest to add a line break before function body even in one-line functions

Reply:
Normally I'd be okay with this change, but for these vctrs definitions I've found it helpful to have them laid out like this when editing them --- it's easier to keep them consistent across the various combinations of definitions that should follow similar patterns. If it's alright without you I'd like to leave them this way in this file (I am adjusting line-spacing everywhere else).